addons/googledrive/models.py

```diff
@@ -0,0 +1,243 @@
+    serializer = GoogleDriveSerializer
+
+
+class NodeSettings(BaseStorageAddon, BaseOAuthNodeSettings):
+    oauth_provider = GoogleDriveProvider
+    provider_name = 'googledrive'
+
+    folder_id = models.TextField(default=None, null=True, blank=True)
+    folder_path = models.TextField(null=True, blank=True)
+    serializer = GoogleDriveSerializer
+    user_settings = models.ForeignKey(UserSettings, null=True, blank=True)
+
+    _api = None
+
+    @property
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider(self.external_account)
+        return self._api
+
+    @property
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+            metadata={'folder': self.folder_id}
+        ))
+
+    @property
+    def folder_name(self):
+        if not self.folder_id:
+            return None
+
+        if self.folder_path != '/':
+            # `urllib` does not properly handle unicode.
+            # encode input to `str`, decode output back to `unicode`
+            return urllib.unquote(os.path.split(self.folder_path)[1].encode('utf-8')).decode('utf-8')
+        else:
+            return '/ (Full Google Drive)'
+
+    def fetch_folder_name(self):
+        return self.folder_name
+
+    def clear_settings(self):
+        self.folder_id = None
+        self.folder_path = None
+
+    def get_folders(self, **kwargs):
+        node = self.owner
+
+        #  Defaults exist when called by the API, but are `None`
+        path = kwargs.get('path') or ''
+        folder_id = kwargs.get('folder_id') or 'root'
+
+        try:
+            access_token = self.fetch_access_token()
+        except exceptions.InvalidAuthError:
+            raise HTTPError(403)
+
+        client = GoogleDriveClient(access_token)
+        if folder_id == 'root':
+            about = client.about()
+
+            return [{
+                'addon': self.config.short_name,
+                'path': '/',
+                'kind': 'folder',
+                'id': about['rootFolderId'],
+                'name': '/ (Full Google Drive)',
+                'urls': {
+                    'folders': api_v2_url('nodes/{}/addons/googledrive/folders/'.format(self.owner._id),
+                        params={
+                            'path': '/',
+                            'id': about['rootFolderId']
+                    })
+                }
+            }]
+
+        contents = [
+            to_hgrid(item, node, path=path)
+            for item in client.folders(folder_id)
+        ]
+        return contents
+
+    def set_folder(self, folder, auth):
+        """Configure this addon to point to a Google Drive folder
+
+        :param dict folder:
+        :param User user:
+        """
+        self.folder_id = folder['id']
```

Comment:
Sometimes it's better to let the `KeyError` blow up if it means someone broke something so bad and they otherwise wouldn't know it. I was just checking.